nixos/modules/services/web-apps/hedgedoc.nix

```diff
@@ -1028,45 +227,91 @@ in
     };
   };
 
-  config = mkIf cfg.enable {
-    assertions = [
-      { assertion = cfg.settings.db == {} -> (
-          cfg.settings.dbURL != "" && cfg.settings.dbURL != null
-        );
-        message = "Database configuration for HedgeDoc missing."; }
-    ];
-    users.groups.${name} = {};
+  config = lib.mkIf cfg.enable {
+    users.groups.${name} = { };
     users.users.${name} = {
       description = "HedgeDoc service user";
       group = name;
-      extraGroups = cfg.groups;
-      home = cfg.workDir;
-      createHome = true;
       isSystemUser = true;
     };
 
+    services.hedgedoc.settings = {
+      defaultNotePath = lib.mkDefault "${cfg.package}/public/default.md";
+      docsPath = lib.mkDefault "${cfg.package}/public/docs";
+      viewPath = lib.mkDefault "${cfg.package}/public/views";
+    };
+
     systemd.services.hedgedoc = {
       description = "HedgeDoc Service";
+      documentation = [ "https://docs.hedgedoc.org/" ];
       wantedBy = [ "multi-user.target" ];
       after = [ "networking.target" ];
-      preStart = ''
-        ${pkgs.envsubst}/bin/envsubst \
-          -o ${cfg.workDir}/config.json \
-          -i ${prettyJSON cfg.settings}
-        mkdir -p ${cfg.settings.uploadsPath}
-      '';
+      preStart =
+        let
+          configFile = settingsFormat.generate "hedgedoc-config.json" {
+            production = cfg.settings;
+          };
+        in
+        ''
+          ${pkgs.envsubst}/bin/envsubst \
+            -o /run/${name}/config.json \
+            -i ${configFile}
+          ${pkgs.coreutils}/bin/mkdir -p ${cfg.settings.uploadsPath}
+        '';
       serviceConfig = {
-        WorkingDirectory = cfg.workDir;
-        StateDirectory = [ cfg.workDir cfg.settings.uploadsPath ];
-        ExecStart = "${lib.getExe cfg.package}";
-        EnvironmentFile = mkIf (cfg.environmentFile != null) [ cfg.environmentFile ];
+        User = name;
+        Group = name;
```

Comment:
Does this service require stable UID and GID or could it be run as `DynamicUser`?

Reply:
While the service doesn't require a stable UID and GID by itself, I've had a hard time trying to let nginx be part of the group of a DynamicUser, in order to be able to talk to its UNIX socket. https://github.com/systemd/systemd/issues/7253

Reply:
Completely untested but maybe something like this, using `systemd-socket-proxyd` to guard access to the private socket.

```
# hedgedoc-sock.socket
[Socket]
ListenStream=/run/hedgedoc.sock
SocketUser=nginx
SocketGroup=nginx

[Install]
WantedBy=sockets.target
```

```
# hedgedoc-sock.service
[Unit]
Description="Forwarding to private Hedgedoc socket"
Requires=hedgedoc.service
After=hedgedoc.service

[Service]
ExecStart=/run/current-system/sw/lib/systemd/systemd-socket-proxyd /path/to/private/hedgedoc.sock
```

Reply:
Are you suggesting we make this a part of the module? If not, I don't think shipping a broken systemd config for the program is acceptable (broken in the sense that `DynamicUser` doesn't let us use unix sockets, which hedgedoc states as a feature). I know it's a bit of a slippery slope argument, but I'm having some feelings against using the socket proxy for every program that supports unix sockets in nixpkgs, just to allow us to use `DynamicUser`.

Maybe we could leave this for another PR? The hedgedoc system user existed before this PR, so I'm not adding or removing anything by reusing the UID and GID here.

Reply:
That's a fair point. The additional layer of indirection doesn't really add anything useful.

Reply:
I'll just mark this as resolved for now then.